### ISSUE TITLE
Ordering::and_then

### ISSUE THREAD

**Issue by digama0** [rust-lang/rust#34774]
This seems like an obvious and missing function from `std::cmp::Ordering`:

```
impl Ordering {
    pub fn then(self, other: Ordering) -> Ordering {
        match self {
            Ordering::Equal => other,
            o => o,
        }
    }

    pub fn and_then<F: FnOnce() -> Ordering>(self, other: F) -> Ordering {
        match self {
            Ordering::Equal => other(),
            o => o,
        }
    }
}
```

This is useful for creating comparators which do lexicographic order. The second variant allows for skipping the order calculation if it is not necessary. Names are of course open to bikeshedding.

**Comment by nagisa** [comment]
Seems like a weird name for such a function IMO. Also, I believe, this should go into the RFC repo.

**Comment by digama0** [comment]
@nagisa Java has a similar function for comparators, called `thenComparing`. The idea is that you can read a comparison function like

    self.name.cmp(other.name).then(self.address.cmp(other.address))

as "compare by name, **then** by address". The `then` / `and_then` difference with a closure is inspired by Rust's function names for `Option` like `or` (by value) / `or_else` (by closure).

I don't know about the RFC thing, I usually see library functions posted here. Plus, I'm not confident I can handle the politics of RFC -> feature -> stabilization, so I am hoping someone more prominent thinks this is a good idea and picks up the baton.

**Comment by nagisa** [comment]
> I don't know about the RFC thing, I usually see library functions posted here. Plus, I'm not confident I can handle the politics of RFC -> feature -> stabilization, so I am hoping someone more prominent thinks this is a good idea and picks up the baton.

You don’t need a full RFC, only this issue should be in *that* repository, since this issue (as well as many other feature requests) are RFCs in their most primitive form.

> The idea is that you can read a comparison function like

When you put it like this it seems to make sense, but I feel like there still may be better ways to call it, especially for `and_then` variation.

**Comment by apasel422** [comment]
This was implemented in #37054 with #37053 as the tracking issue.
